### PR TITLE
Beginning of documentation

### PR DESCRIPTION
**Pull request by Mironor**
This is just a proposition and by no means something final.

Why I propose split huge files into small files in a package:
pros:
+ It's recommended by Scala style guide http://docs.scala-lang.org/style/files.html
+ It's easier to make changes as you will be concentrated only on one file at a time
+ It's better for version control

cons:
- Breaks perfect backward compatibility (will need to change import statements)

I also added few comments on things I already know how to use.

This is just a small pull request that is not intended to be merged. Instead it is here to make sure I'm moving in a right direction as soon as possible (and keep track on progress)

### CODE REVIEW

**Comment by OlegIlyenko** [comment]
Great, thanks a lot! I will look at it as soon as I get an opportunity (I think tomorrow)

**Comment by OlegIlyenko** [comment]
I am sorry to have kept you waiting, and thanks again for this PR! 

First about the split of `Injectable.scala`. For me personally it always was a single logical unit. `Injectable` and `OpenInjectable` (which make most of this file) are highly coupled together and change to one often result in change of another. Other classes in there are just DSL support classes which all together represent this feature. By looking in this file you can immediately see all of the classes that implement injection feature, which I find nice. I would say, that following clause from Scala style guide apply here:

> Another case is when multiple classes logically form a single, cohesive group, sharing concepts to the point where maintenance is greatly served by containing them within a single file.

Because of these factors, I think I would prefer to keep then in one file. That said, I still agree with you. In many cases it's actually advantageous to keep compilation units in separate files. I just feel that `Injectable.scala` is not very good example of it. I think `Module.scala`, for instance, would be a better candidate for the split. It has a lot of implementations of `Injector` trait, but still most of them are pretty independent.

The package change you introduced will indeed break backwards compatibility. For pretty long time now I tried to keep scaldi stable and backwards-compatible because more and more people rely on it now. So I feel that it is important to reduce incompatible changes to minimum. The thing with `Injectable` is that it will affect all of the users without exceptions, since this is such a fundamental feature. If it was something with less impact, then maybe we could sneak it in :)

That said, we can always go though the deprecation process and define some aliases in `scaldi` package which will point to `scaldi.injectable` package. But I think it touches a bigger topic of version 1.0 :) I was thinking about this for a while now. I feel that it's about the right time to release the version 1.0 bested on different factors (for the most part user feedback and recent integration with Play 2.4). If we decide to restructure the library, then I think that we still need to go through deprecation process with series of milestone releases of 1.0. Maybe this Autumn would be a good time to do it, not sure about the exact dates yet though.

About the structure itself... I think for me it was always important that `import scaldi._` will import the most important and widely used functionality. It still makes it possible to define classes in dedicated packages and then maybe define type aliases or some implementations in the `scaldi` package (I guess we can take some of the inspiration from different libraries like scala standard library itself, scalaz, etc.). I think it will strike a nice balance between user convenience and and library structure for better maintenance. That said, since scaldi is such a small library, I would try to avoid putting too much structure into it because I feel that it can unnecessary complicate things. Still some structure can help organize things, for example we can introduce `scaldi.binding` and `scaldi.injector` which can nicely group related classes.

Hmm, it was a looong write :) I hope it clarified a bit my vision on the library and it's future. I tried to described my honest opinion here, but please don't be discouraged if you disagree with some of the things I said. I'm very open to new ideas, so I would be happy to hear you opinion on this and learn from it!

**Comment by Mironor** [comment]
Nice and very informal reply, thanks : )

I'll rollback the package change, flat architecture is ok. 
I'll suggest an extraction of a class in a separate file (in the same package) if I see a suitable one.
I'll do a javadoc for methods/class as far as my understanding of docs/tests goes on my own pace, please come to see this pull request from time to time.

We'll wait for a major version before considering a backward incompatibility, you are absolutely right!

**Comment by Mironor** [comment]
I covered Injectable with javadoc, gave small examples for the most used methods.
I have a small problem with defining the purpose of injectWithConstructorDefault method, is the definition in javadoc ok?
Please tell me if the style of javadoc is ok for you.

File: src/main/scala/scaldi/Injectable.scala [hunk 1]
```diff
@@ -84,37 +204,112 @@ trait OpenInjectable extends Injectable {
...
+   */
   def which(by: ByWord) = and(by)
+
+  /**
+   * Alias to "and"
+   * Example:
+   * inject [Database] ('remote that by default defaultDb),
+   *
+   * @param by ByWord that follows the "that"
+   * @return
```
Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: I think we can avoid empty `@return`

File: src/main/scala/scaldi/Injectable.scala [hunk 2]
```diff
@@ -1,67 +1,193 @@
...
+ */
+trait Injectable extends Wire {
+  /**
+   * Injects a dependency that was previously defined in a module
+   * Example:
+   * val database = inject[Database]  // database can now be used as an instance of Database class
+   *
+   * @param injector implicit, should be defined in the scope
+   * @param tt implicit, already defined in the scope
+   * @param nn implicit, already defined in the scope
```
Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: I would expand a bit on all of these implicits and explain their purpose. Maybe something like this:

* `injector` - implicit `Injector`, which should available in the scope. It's used to lookup bindings.
* `tt` - ensures that meta-information if type `T` is available at the runtime
* `nn` - ensures, that type `T` is not inferred as `Nothing` by scala compiler, since `inject` needs to know the actual type at the runtime.

File: src/main/scala/scaldi/Injectable.scala [hunk 3]
```diff
@@ -1,67 +1,193 @@
...
+ * It’s important to understand, that it’s the only purpose of it. So it is
+ * completely stateless and knows nothing about actual bindings you have
+ * defined in the module.
+ *
+ * In order to actually find and inject dependencies, "inject" method always
+ * takes an implicit parameter of type Injector.
+ */
+trait Injectable extends Wire {
+  /**
+   * Injects a dependency that was previously defined in a module
```
Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: I think in this and other places we can put an empty line to separate the paragraphs. Otherwise scaladoc will render  everything in one line.

File: src/main/scala/scaldi/Injectable.scala [hunk 4]
```diff
@@ -1,67 +1,193 @@
...
+   * Example:
+   * val database = inject[Database]  // database can now be used as an instance of Database class
```
Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: for the source code snippets I would prefer to use `` `some code` `` (back-tick) or 
```
{{{
multi
line 
code
}}}
```

File: src/main/scala/scaldi/Injectable.scala [hunk 5]
```diff
@@ -1,67 +1,193 @@
...
-  // in case is identifier goes at first
-
-  protected implicit def canBeIdentifiedToConstraints[T : CanBeIdentifier](target: T) =
+  /**
+   * Implicit cast of classes that can be identifiers to an InjectConstraints with the identified as a first element
+   * Often used when a list of identifiers should be passed (for example for "inject(constraint)" method)
+   * This lets us to put the identifier in a first place to make the parameters look readable
+   *
+   * @param target Instance of a class that qualifies as an identifier
+   * @tparam T TypeClass of a class that qualifies as an identifier
```
Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: I find wording here a bit confusing. What do you mean by `TypeClass`? In my understanding `CanBeIdentifier` is a type-class in this case and `T` is just some type that can be treated as an identifier (there is an implicit instance of `CanBeIdentifier` defined somewhere for type `T`)

File: src/main/scala/scaldi/Injectable.scala [hunk 6]
```diff
@@ -84,37 +204,112 @@ trait OpenInjectable extends Injectable {
...
+  override implicit def canBeIdentifiedToConstraints[T: CanBeIdentifier](target: T): InjectConstraints[Nothing] =
+    super.canBeIdentifiedToConstraints[T](target)
+
+  override val identified = new IdentifiedWord
+
+  override val by = new ByWord
 }
 
+/**
+ * Injectable trait is completely stateless, so it can be declared as an Object
```
Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: Maybe we can provide an example of it's usage. Something like 

```scala
import scaldi.Injectable._

val database = inject [Database]
```

Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: Also maybe we can rephrase it a bit. For example like this: 

Companion object of `Injectiable` which can be just imported in contrast to `Injectable` trait which always need to be extended. After importing it, you can use an injection DSL without extending the trait.

File: src/main/scala/scaldi/Injectable.scala [hunk 7]
```diff
@@ -1,67 +1,193 @@
...
-    () => inject(injector, tt, nn)
 
-  protected def injectProvider[T](constraints: => InjectConstraints[T])
-                                 (implicit injector: Injector, tt: TypeTag[T], nn: NotNothing[T]): () => T =
-    () => inject(constraints)(injector, tt, nn)
+import scala.language.{implicitConversions, postfixOps}
+import scala.reflect.runtime.universe.TypeTag
 
+/**
+ * The only responsibility of this trait is to provide you with inject function
```
Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: Maybe we can start this part with small and gentle introduction. Maybe something like: 

`Injectable` trait provides and injection DSL. You can extend in your classes in order use `inject` methods.

Also, I think it can be helpful to provide some basic example at the end of this scaladoc

File: src/main/scala/scaldi/Injectable.scala (hunk 6 above)
Comment (Mironor) [inline src/main/scala/scaldi/Injectable.scala]: Great! I didn't know it was possible (I wondered how to inject things in traits, this seems to be the answer)

File: src/main/scala/scaldi/Injectable.scala [hunk 8]
```diff
@@ -1,67 +1,193 @@
...
+   * @tparam T class of the injected dependency
+   * @return injection defined by identifiers or default value if injection is not found
+   */
   protected def injectWithDefault[T](injector: Injector, default: => T)(ids: List[Identifier]) =
     injector getBinding ids flatMap (_.get) map (_.asInstanceOf[T]) getOrElse default
 
+  /**
+   * ??? Injects a dependency defined by TypeClass. If dependency is not found, initializes it with constructor
+   * Constructor must have a parameter paramName with a default value
+   * Uses reflection
```
Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: This methods is actually more or less part of internal implementation of `injected` macro, so I don't expect people to use it directly. I think I would rephrase it like this:

```
Injects a dependency defined by type `T`. If dependency is not found, initializes it with the default value from the constructor of type `T`.

This implies, that constructor must have one constructor with parameter `paramName` which has a default value.

Method is part of `injected` macro implementation and not intended to be used outside of it.

Uses reflection.
```

**Comment by OlegIlyenko** [comment]
@Mironor thank you very much for the effort you put in this PR! Highly appreciated!

In general I like it. I put a few comments for some more specific things :)

File: src/main/scala/scaldi/Injectable.scala (hunk 8 above)
Comment (Mironor) [inline src/main/scala/scaldi/Injectable.scala]: How does this macro implementation work? I didn't find any usage of this method aside from in OpenInjectable.
And if it's not intentended to be used directly, should we remove it from OpenInjectable and made private in this case

Comment (OlegIlyenko) [inline src/main/scala/scaldi/Injectable.scala]: The usage is generated in the macro: https://github.com/scaldi/scaldi/blob/master/src/main/scala/scaldi/Wire.scala#L66

You are right. It was a rush decision of mine to put it in the `Injectable`. I think some other, more isolated and less visible, place would be more suitable for it. I guess a good candidate for the next version :) https://github.com/scaldi/scaldi/issues/55

**Comment by OlegIlyenko** [comment]
Thank for adding the improvements I suggested!

**Comment by OlegIlyenko** [comment]
I just made a small improvements, hope you don't mind:

https://github.com/scaldi/scaldi/commit/03d4824bd4efd6be9f30722820a05e79a4fa1caf

Now it's your time to do the review :)

I also published the docs, so they are already online:

http://scaldi.github.io/scaldi/latest/api/index.html#scaldi.Injectable$
